Sources/SourceKitLSP/Swift/CodeActions/PackageManifestEdits.swift

```diff
@@ -0,0 +1,254 @@
+          edit: edits.asWorkspaceEdit(snapshot: scope.snapshot)
+        )
+      ]
+    } catch {
+      return []
+    }
+  }
+
+  /// A list of target kinds that allow the creation of tests.
+  static let targetsThatAllowTests: Set<String> = [
+    "executableTarget",
+    "macro",
+    "target",
+  ]
+
+  /// Produce code actions to add a product if we are currently on
+  /// a target for which we can create a product.
+  static func maybeAddProductActions(
+    call: FunctionCallExprSyntax,
+    in scope: SyntaxCodeActionScope
+  ) -> [CodeAction] {
+    guard let calledMember = call.findMemberAccessCallee(),
+      targetsThatAllowProducts.contains(calledMember),
+      let targetName = call.findStringArgument(label: "name")
+    else {
+      return []
+    }
+
+    do {
+      let type: ProductType =
+        calledMember == "executableTarget"
+        ? .executable
+        : .library(.automatic)
+
+      // Describe the target we are going to create.
+      let product = try ProductDescription(
+        name: "\(targetName)",
+        type: type,
+        targets: [targetName]
+      )
+
+      let edits = try AddProduct.addProduct(product, to: scope.file)
+      return [
+        CodeAction(
+          title: "Add product to export this target",
+          kind: .refactor,
+          edit: edits.asWorkspaceEdit(snapshot: scope.snapshot)
+        )
+      ]
+    } catch {
+      return []
+    }
+  }
+
+  /// A list of target kinds that allow the creation of tests.
+  static let targetsThatAllowProducts: Set<String> = [
+    "executableTarget",
+    "target",
+  ]
+}
+
+fileprivate extension PackageEditResult {
+  /// Translate package manifest edits into a workspace edit. This can
+  /// involve both modifications to the manifest file as well as the creation
+  /// of new files.
+  func asWorkspaceEdit(snapshot: DocumentSnapshot) -> WorkspaceEdit {
+    // The edits to perform on the manifest itself.
+    let manifestTextEdits = manifestEdits.map { edit in
+      TextEdit(
+        range: snapshot.range(of: edit.range),
+        newText: edit.replacement
+      )
+    }
+
+    // If we couldn't figure out the manifest directory, or there are no
+    // files to add, the only changes are the manifest edits. We're done
+    // here.
+    let manifestDirectoryURI = snapshot.uri.fileURL?
+      .deletingLastPathComponent()
+    guard let manifestDirectoryURI, !auxiliaryFiles.isEmpty else {
+      return WorkspaceEdit(
+        changes: [snapshot.uri: manifestTextEdits]
+      )
+    }
+
+    // Use the more full-featured documentChanges, which takes precedence
+    // over the individual changes to documents.
+    var documentChanges: [WorkspaceEditDocumentChange] = []
+
+    // Put the manifest changes into the array.
+    documentChanges.append(
+      .textDocumentEdit(
+        .init(
+          textDocument: .init(snapshot.uri, version: nil),
+          edits: manifestTextEdits.map { .textEdit($0) }
+        )
```

Comment:
Since we have an open snapshot for `Package.swift`, we should also send the version. From the LSP spec

> If an optional versioned text document identifier is sent from the server to the client and the file is not open in the editor (the server has not received an open notification before) the server can send `null` to indicate that the version is known and the content on disk is the master (as specified with document content ownership).

Which I read like you should only specify `nil` as the `version` if the file isn’t open in the editor, but in this case it is.

Also, nitpick: I generally dislike `.init` because while it’s easy to write, it’s harder to read, not knowing which type gets created.

```suggestion
        TextDocumentEdit(
          textDocument: OptionalVersionedTextDocumentIdentifier(snapshot.uri, version: snapshot.version),
          edits: manifestTextEdits.map { .textEdit($0) }
        )
```

Reply:
I agree on the `TextDocumentEdit`, but `OptionalVersionedTextDocumentIdentifier`... yuck.